File: services/rpi-service/protocol.py
"""Tüm katmanlar arası ortak mesaj sözleşmesi.

WebSocket (tarayıcı), ZMQ (PC) ve iç event-bus üzerinden geçen JSON mesajların
şemasını tek yerde tutar. Frontend `src/lib/protocol.ts` dosyası bu yapıyla
birebir aynı olmalıdır."""
from __future__ import annotations
from dataclasses import dataclass, asdict, field
from typing import Literal, Optional
import time

# ───────────────────────────────────────────────────────────────────────
# İstemciden (UI) gelen komutlar
# ───────────────────────────────────────────────────────────────────────
Mode = Literal["MANUEL", "OTONOM", "SURU"]
Threat = Literal["DOST", "DUSMAN", "BILINMEYEN"]
Status = Literal["HAZIR", "TARAMA", "KILITLEME", "ATIS", "YENIDEN_YUKLE", "DURDURULDU"]
LogLevel = Literal["INFO", "WARN", "ERROR", "SUCCESS"]


@dataclass
class Command:
    action: str
    mode: Optional[Mode] = None
    pan: Optional[float] = None
    tilt: Optional[float] = None
    target_id: Optional[str] = None
    threat: Optional[Threat] = None
    safety: Optional[bool] = None
    steps: Optional[int] = None


# ───────────────────────────────────────────────────────────────────────
# Sunucudan (RPi) yayınlanan mesajlar
# ───────────────────────────────────────────────────────────────────────
@dataclass
class Motors:
    pan: float = 0.0
    tilt: float = 0.0
    magazine: float = 0.0


@dataclass
class Sensors:
    lidar_distance: float = 0.0
    lidar_valid: bool = False
    fps: float = 0.0
    latency_ms: float = 0.0


@dataclass
class Target:
    id: str
    label: str
    type: Threat
    x: float          # 0..1 normalize görüntü piksel
    y: float
    distance: float   # metre
    speed: float      # km/h
    heading: float    # derece
    confidence: float # 0..100
    priority: int     # 0..10
    locked: bool = False


@dataclass
class Telemetry:
    type: str = "telemetry"
    ts: float = field(default_factory=time.time)
    mode: Mode = "MANUEL"
    status: Status = "HAZIR"
    safety_on: bool = True
    emergency_stop: bool = False
    ammo: int = 24
    max_ammo: int = 24
    shots_fired: int = 0
    motors: Motors = field(default_factory=Motors)
    sensors: Sensors = field(default_factory=Sensors)


@dataclass
class TargetsMsg:
    type: str = "targets"
    ts: float = field(default_factory=time.time)
    targets: list[Target] = field(default_factory=list)
    locked_id: Optional[str] = None


@dataclass
class LogMsg:
    type: str = "log"
    ts: float = field(default_factory=time.time)
    level: LogLevel = "INFO"
    message: str = ""


@dataclass
class AckMsg:
    type: str = "ack"
    action: str = ""
    ok: bool = True
    reason: Optional[str] = None
    ts: float = field(default_factory=time.time)


def to_dict(obj) -> dict:
    """dataclass → JSON-serializable dict (nested dataclass'ları da çevirir)."""
    return asdict(obj)

File: artifacts/savunma-kontrol/src/lib/protocol.ts
// Sunucu (RPi Python servisi) ile birebir aynı mesaj sözleşmesi.
// Bu dosya `services/rpi-service/protocol.py` ile senkron tutulmalıdır.

import type { SystemMode, ThreatType, Target as UITarget, LogEntry } from "./system-state";

// ───── Komutlar (UI → server) ────────────────────────────────────────
export type CommandAction =
  | "set_mode"
  | "emergency_stop"
  | "reset"
  | "set_motors"
  | "magazine_step"
  | "home_motors"
  | "lock_target"
  | "classify_target"
  | "fire"
  | "set_safety"
  | "ping";

export interface Command {
  action: CommandAction;
  mode?: SystemMode;
  pan?: number;
  tilt?: number;
  target_id?: string | null;
  threat?: ThreatType;
  safety?: boolean;
  steps?: number;
}

// ───── Sunucudan gelen mesajlar (server → UI) ────────────────────────
export interface ServerTelemetry {
  type: "telemetry";
  ts: number;
  mode: SystemMode;
  status: string;
  safety_on: boolean;
  emergency_stop: boolean;
  ammo: number;
  max_ammo: number;
  shots_fired: number;
  motors: { pan: number; tilt: number; magazine: number };
  sensors: {
    lidar_distance: number;
    lidar_valid: boolean;
    fps: number;
    latency_ms: number;
  };
}

export interface ServerTarget {
  id: string;
  label: string;
  type: ThreatType;
  x: number;
  y: number;
  distance: number;
  speed: number;
  heading: number;
  confidence: number;
  priority: number;
  locked: boolean;
}

export interface ServerTargetsMsg {
  type: "targets";
  ts: number;
  targets: ServerTarget[];
  locked_id: string | null;
}

export interface ServerLogMsg {
  type: "log";
  ts: number;
  level: LogEntry["level"];
  message: string;
}

export interface ServerAck {
  type: "ack";
  action: string;
  ok: boolean;
  reason?: string;
  ts: number;
}

export type ServerMessage = ServerTelemetry | ServerTargetsMsg | ServerLogMsg | ServerAck;

// ───── Sunucu hedefini UI hedefine çevir ─────────────────────────────
export function serverTargetToUI(t: ServerTarget): UITarget {
  return {
    id: t.id,
    label: t.label,
    type: t.type,
    x: t.x,
    y: t.y,
    distance: t.distance,
    confidence: t.confidence,
    speed: t.speed,
    heading: t.heading,
    locked: t.locked,
    priority: t.priority,
  };
}
